File: contrib/c/dynet_c/softmax-builder.cc
#include <dynet_c/config.h>

#include <dynet/cfsm-builder.h>
#include <dynet_c/internal.h>
#include <dynet_c/softmax-builder.h>

#include <vector>

using dynet_c::internal::to_c_ptr;
using dynet_c::internal::to_cpp_ptr;
using dynet_c::internal::to_c_ptr_from_value;

DYNET_C_STATUS dynetDeleteSoftmaxBuilder(dynetSoftmaxBuilder_t *builder) try {
  DYNET_C_CHECK_NOT_NULL(builder);
  delete to_cpp_ptr(builder);
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

DYNET_C_STATUS dynetResetSoftmaxBuilderGraph(
    dynetSoftmaxBuilder_t *builder, dynetComputationGraph_t *cg,
    DYNET_C_BOOL update) try {
  DYNET_C_CHECK_NOT_NULL(builder);
  DYNET_C_CHECK_NOT_NULL(cg);
  to_cpp_ptr(builder)->new_graph(*to_cpp_ptr(cg), update);
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

DYNET_C_STATUS dynetApplySoftmaxBuilderNegLogSoftmaxOne(
    dynetSoftmaxBuilder_t *builder, const dynetExpression_t *rep,
    uint32_t classidx, dynetExpression_t **newobj) try {
  DYNET_C_CHECK_NOT_NULL(builder);
  DYNET_C_CHECK_NOT_NULL(rep);
  DYNET_C_CHECK_NOT_NULL(newobj);
  *newobj = to_c_ptr_from_value(
      to_cpp_ptr(builder)->neg_log_softmax(*to_cpp_ptr(rep), classidx));
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

DYNET_C_STATUS dynetApplySoftmaxBuilderNegLogSoftmax(
    dynetSoftmaxBuilder_t *builder, const dynetExpression_t *rep,
    const uint32_t *classidxs, size_t n, dynetExpression_t **newobj) try {
  DYNET_C_CHECK_NOT_NULL(builder);
  DYNET_C_CHECK_NOT_NULL(rep);
  DYNET_C_CHECK_NOT_NULL(classidxs);
  DYNET_C_CHECK_NOT_NULL(newobj);
  *newobj = to_c_ptr_from_value(
      to_cpp_ptr(builder)->neg_log_softmax(
          *to_cpp_ptr(rep), std::vector<uint32_t>(classidxs, classidxs + n)));
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

DYNET_C_STATUS dynetSampleFromSoftmaxBuilder(
    dynetSoftmaxBuilder_t *builder, const dynetExpression_t *rep,
    uint32_t *retval) try {
  DYNET_C_CHECK_NOT_NULL(builder);
  DYNET_C_CHECK_NOT_NULL(rep);
  DYNET_C_CHECK_NOT_NULL(retval);
  *retval = to_cpp_ptr(builder)->sample(*to_cpp_ptr(rep));
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

DYNET_C_STATUS dynetGetSoftmaxBuilderFullLogDistribution(
    dynetSoftmaxBuilder_t *builder, const dynetExpression_t *rep,
    dynetExpression_t **newobj) try {
  DYNET_C_CHECK_NOT_NULL(builder);
  DYNET_C_CHECK_NOT_NULL(rep);
  DYNET_C_CHECK_NOT_NULL(newobj);
  *newobj = to_c_ptr_from_value(
      to_cpp_ptr(builder)->full_log_distribution(*to_cpp_ptr(rep)));
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

DYNET_C_STATUS dynetGetSoftmaxBuilderFullLogits(
    dynetSoftmaxBuilder_t *builder, const dynetExpression_t *rep,
    dynetExpression_t **newobj) try {
  DYNET_C_CHECK_NOT_NULL(builder);
  DYNET_C_CHECK_NOT_NULL(rep);
  DYNET_C_CHECK_NOT_NULL(newobj);
  *newobj = to_c_ptr_from_value(
      to_cpp_ptr(builder)->full_logits(*to_cpp_ptr(rep)));
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

DYNET_C_STATUS dynetGetSoftmaxBuilderParameterCollection(
    dynetSoftmaxBuilder_t *builder, dynetParameterCollection_t **newobj) try {
  DYNET_C_CHECK_NOT_NULL(builder);
  DYNET_C_CHECK_NOT_NULL(newobj);
  *newobj = to_c_ptr(&to_cpp_ptr(builder)->get_parameter_collection());
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

DYNET_C_STATUS dynetCreateStandardSoftmaxBuilder(
    uint32_t rep_dim, uint32_t num_classes, dynetParameterCollection_t *pc,
    DYNET_C_BOOL bias, dynetSoftmaxBuilder_t **newobj) try {
  DYNET_C_CHECK_NOT_NULL(pc);
  DYNET_C_CHECK_NOT_NULL(newobj);
  *newobj = to_c_ptr(new dynet::StandardSoftmaxBuilder(
      rep_dim, num_classes, *to_cpp_ptr(pc), bias));
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

DYNET_C_STATUS dynetCreateStandardSoftmaxBuilderFromParameters(
    dynetParameter_t *p_w, dynetParameter_t *p_b,
    dynetSoftmaxBuilder_t **newobj) try {
  DYNET_C_CHECK_NOT_NULL(p_w);
  DYNET_C_CHECK_NOT_NULL(newobj);
  if (p_b) {
    *newobj = to_c_ptr(
        new dynet::StandardSoftmaxBuilder(*to_cpp_ptr(p_w), *to_cpp_ptr(p_b)));
  } else {
    *newobj = to_c_ptr(new dynet::StandardSoftmaxBuilder(*to_cpp_ptr(p_w)));
  }
  return DYNET_C_OK;
} DYNET_C_HANDLE_EXCEPTIONS

File: contrib/rust/src/softmax_builder.rs
use std::ptr::{self, NonNull};

use dynet_sys;

use super::{
    ApiResult, ComputationGraph, Expression, Parameter, ParameterCollection, Result, Wrap,
};

/// `SoftmaxBuilder` trait
pub trait SoftmaxBuilder: Wrap<dynet_sys::dynetSoftmaxBuilder_t> {
    /// Initializes the parameters in the computation graph.
    fn new_graph(&mut self, cg: &mut ComputationGraph, update: bool) {
        unsafe {
            check_api_status!(dynet_sys::dynetResetSoftmaxBuilderGraph(
                self.as_mut_ptr(),
                cg.as_mut_ptr(),
                update as u32,
            ));
        }
    }

    /// Computes negative log probability of a class.
    fn neg_log_softmax_one<E: AsRef<Expression>>(&mut self, rep: E, classidx: u32) -> Expression {
        unsafe {
            let mut expr_ptr: *mut dynet_sys::dynetExpression_t = ptr::null_mut();
            check_api_status!(dynet_sys::dynetApplySoftmaxBuilderNegLogSoftmaxOne(
                self.as_mut_ptr(),
                rep.as_ref().as_ptr(),
                classidx,
                &mut expr_ptr,
            ));
            Expression::from_raw(expr_ptr, true)
        }
    }

    /// Computes batched negative log probability of a class.
    fn neg_log_softmax<E: AsRef<Expression>>(&mut self, rep: E, classidxs: &[u32]) -> Expression {
        unsafe {
            let mut expr_ptr: *mut dynet_sys::dynetExpression_t = ptr::null_mut();
            check_api_status!(dynet_sys::dynetApplySoftmaxBuilderNegLogSoftmax(
                self.as_mut_ptr(),
                rep.as_ref().as_ptr(),
                classidxs.as_ptr(),
                classidxs.len(),
                &mut expr_ptr,
            ));
            Expression::from_raw(expr_ptr, true)
        }
    }

    /// Samples from the softmax distribution.
    fn sample<E: AsRef<Expression>>(&mut self, rep: E) -> u32 {
        unsafe {
            let mut retval: u32 = 0;
            check_api_status!(dynet_sys::dynetSampleFromSoftmaxBuilder(
                self.as_mut_ptr(),
                rep.as_ref().as_ptr(),
                &mut retval,
            ));
            retval
        }
    }

    /// Returns an Expression representing a vector the size of the number of classes.
    fn full_log_distribution<E: AsRef<Expression>>(&mut self, rep: E) -> Expression {
        unsafe {
            let mut expr_ptr: *mut dynet_sys::dynetExpression_t = ptr::null_mut();
            check_api_status!(dynet_sys::dynetGetSoftmaxBuilderFullLogDistribution(
                self.as_mut_ptr(),
                rep.as_ref().as_ptr(),
                &mut expr_ptr,
            ));
            Expression::from_raw(expr_ptr, true)
        }
    }

    /// Returns the logits (before application of the softmax).
    fn full_logits<E: AsRef<Expression>>(&mut self, rep: E) -> Expression {
        unsafe {
            let mut expr_ptr: *mut dynet_sys::dynetExpression_t = ptr::null_mut();
            check_api_status!(dynet_sys::dynetGetSoftmaxBuilderFullLogits(
                self.as_mut_ptr(),
                rep.as_ref().as_ptr(),
                &mut expr_ptr,
            ));
            Expression::from_raw(expr_ptr, true)
        }
    }

    /// Gets the ParameterCollection containing the softmax parameters.
    fn get_parameter_collection(&mut self) -> ParameterCollection {
        unsafe {
            let mut pc_ptr: *mut dynet_sys::dynetParameterCollection_t = ptr::null_mut();
            check_api_status!(dynet_sys::dynetGetSoftmaxBuilderParameterCollection(
                self.as_mut_ptr(),
                &mut pc_ptr,
            ));
            ParameterCollection::from_raw(pc_ptr, false)
        }
    }
}

macro_rules! impl_softmax_builder {
    ($name:ident) => {
        impl_wrap_owned!($name, dynetSoftmaxBuilder_t);
        impl_drop!($name, dynetDeleteSoftmaxBuilder);
        impl SoftmaxBuilder for $name {}
    };
}

/// A builder for the standard softmax.
#[derive(Debug)]
pub struct StandardSoftmaxBuilder {
    inner: NonNull<dynet_sys::dynetSoftmaxBuilder_t>,
}

impl_softmax_builder!(StandardSoftmaxBuilder);

impl StandardSoftmaxBuilder {
    /// Creates a new `StandardSoftmaxBuilder`.
    ///
    /// # Arguments
    ///
    /// * rep_dim - Dimension of the input vectors.
    /// * num_classes - Number of classes.
    /// * pc - Parameter collection.
    /// * bias - Whether to use a bias vector or not.
    pub fn new(
        rep_dim: u32,
        num_classes: u32,
        pc: &mut ParameterCollection,
        bias: bool,
    ) -> StandardSoftmaxBuilder {
        unsafe {
            let mut builder_ptr: *mut dynet_sys::dynetSoftmaxBuilder_t = ptr::null_mut();
            check_api_status!(dynet_sys::dynetCreateStandardSoftmaxBuilder(
                rep_dim,
                num_classes,
                pc.as_mut_ptr(),
                bias as u32,
                &mut builder_ptr,
            ));
            StandardSoftmaxBuilder::from_raw(builder_ptr, true)
        }
    }

    /// Creates a new `StandardSoftmaxBuilder` from parameters.
    pub fn from_parameters(p_w: &mut Parameter, p_b: &mut Parameter) -> StandardSoftmaxBuilder {
        unsafe {
            let mut builder_ptr: *mut dynet_sys::dynetSoftmaxBuilder_t = ptr::null_mut();
            check_api_status!(dynet_sys::dynetCreateStandardSoftmaxBuilderFromParameters(
                p_w.as_mut_ptr(),
                p_b.as_mut_ptr(),
                &mut builder_ptr,
            ));
            StandardSoftmaxBuilder::from_raw(builder_ptr, true)
        }
    }
}
